gateway/engine/vertx-shareddata/src/main/java/io/apiman/gateway/engine/vertx/shareddata/SharedGlobalDataRegistry.java

```diff
@@ -0,0 +1,283 @@
+                        } else {
+                            objectMap.putIfAbsent(clientIndex, client, replaceClientFuture.completer());
+                        }
+                        return replaceClientFuture;
+                    // Remove old API key reference
+                    }).compose(oldClientRaw -> {
+                        Client oldClient = (Client) oldClientRaw;
+                        if (oldClientRaw != null && !oldClient.getApiKey().equals(client.getApiKey())) {
+                            objectMap.remove(oldClient.getApiKey(), endFuture.completer());
+                        } else {
+                            endFuture.complete();
+                        }
+                    }, endFuture)
+                    // When finished, call this handler and then resultHandler
+                    .setHandler(handleResult(resultHandler));
+                }
+            } else {
+                resultHandler.handle(AsyncResultImpl.create(compositeResult.cause()));
+            }
+        });
+    }
+
+    @Override
+    public void unregisterClient(Client client, IAsyncResultHandler<Void> resultHandler) {
+        String clientIndex = getClientIndex(client);
+        objectMap.get(clientIndex, handleSuccessfulResult(resultHandler, oldClientRaw -> {
+            Client oldClient = (Client) oldClientRaw;
+            if (oldClient == null) {
+                Exception ex = new RegistrationException(Messages.i18n.format("InMemoryRegistry.ClientNotFound"));
+                resultHandler.handle(AsyncResultImpl.create(ex));
+            } else {
+                Future<Object> future1 = Future.future();
+                Future<Object> future2 = Future.future();
+
+                objectMap.remove(clientIndex, future1.completer());
+                objectMap.remove(oldClient.getApiKey(), future2.completer());
+
+                CompositeFuture.all(future1, future2).setHandler(handleCompositeResult(resultHandler));
+            }
+        }));
+    }
+
+    @Override
+    public void getApi(String organizationId, String apiId, String apiVersion, IAsyncResultHandler<Api> handler) {
+        objectMap.get(getApiIndex(organizationId, apiId, apiVersion), handleResult(handler));
+    }
+
+    @Override
+    public void getClient(String apiKey, IAsyncResultHandler<Client> handler) {
+        objectMap.get(apiKey, handleResult(handler));
+    }
+
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    @Override
+    public void getContract(String apiOrganizationId, String apiId, String apiVersion, String apiKey, IAsyncResultHandler<ApiContract> handler) {
+        String apiIndex = getApiIndex(apiOrganizationId, apiId, apiVersion);
+
+        Future apiFuture = Future.future();
+        Future clientFuture = Future.future();
+
+        objectMap.get(apiIndex, apiFuture.completer());
+        objectMap.get(apiKey, clientFuture.completer());
+
+        CompositeFuture.all(apiFuture, clientFuture).setHandler(compositeResult -> {
```

Comment:
This is cool code.  Took me a couple seconds to grok it, since I haven't yet done much coding with Futures.  But it's pretty nice (for async code). :)